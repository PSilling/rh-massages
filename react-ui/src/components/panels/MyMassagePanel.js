// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Row, Col, Card, Button, CardHeader, CardBody, CardText } from "reactstrap";
import moment from "moment";

// component imports
import CalendarButton from "../iconbuttons/CalendarButton";
import ConfirmationModal from "../modals/ConfirmationModal";

// util imports
import Util from "../../util/Util";
import _t from "../../util/Translations";

/**
 * Massage information panel for My Massages view.
 */
class MyMassagePanel extends Component {
  state = { active: false };

  shouldComponentUpdate(nextProps, nextState) {
    return (
      this.props.type !== nextProps.type ||
      this.props.massage.masseuse !== nextProps.massage.masseuse ||
      this.props.massage.date !== nextProps.massage.date ||
      this.props.massage.ending !== nextProps.massage.ending ||
      this.state.active !== nextState.active
    );
  }

  handleToggle = () => {
    this.setState(prevStata => ({ active: !prevStata.active }));
  };

  cancelMassage = () => {
    Util.put(
      Util.MASSAGES_URL,
      [
        {
          id: this.props.massage.id,
          date: this.props.massage.date,
          ending: this.props.massage.ending,
          masseuse: this.props.massage.masseuse,
          client: null,
          facility: this.props.massage.facility
        }
      ],
      this.props.getCallback
    );
  };

  render() {
    return (
      <div className="mb-3">
        <Row>
          <Col md="3">
            <Card color={this.props.type} outline>
              <CardHeader>
                {moment(this.props.massage.date).format("ddd L")}
                <Button
                  color="light"
                  className="close"
                  aria-label="Close"
                  onClick={this.handleToggle}
                  disabled={this.props.disabled}
                >
                  {!this.props.disabled && <span aria-hidden="true">&times;</span>}
                </Button>
              </CardHeader>
              <CardBody>
                <CardText>{`${_t.translate("Facility")}: ${this.props.massage.facility.name}`}</CardText>
                <CardText>{`${_t.translate("Masseur/Masseuse")}: ${this.props.massage.masseuse}`}</CardText>
                <CardText>
                  {`${_t.translate("Time")}: ${moment(this.props.massage.date).format("HH:mm")}â€“${moment(
                    this.props.massage.ending
                  ).format("HH:mm")}`}
                </CardText>
                <CardText>
                  {`${_t.translate("Event")}: `}
                  <CalendarButton link={Util.getEventLink(this.props.massage)} />
                </CardText>
              </CardBody>
            </Card>
          </Col>
        </Row>
        {this.state.active && (
          <ConfirmationModal
            message={_t.translate("Are you sure you want to unassign yourself from this massage?")}
            onClose={this.handleToggle}
            onConfirm={() => {
              this.handleToggle();
              this.cancelMassage();
            }}
          />
        )}
      </div>
    );
  }
}

MyMassagePanel.propTypes = {
  /** Massage to be printed inside the panel */
  massage: PropTypes.shape({
    id: PropTypes.number,
    masseuse: PropTypes.string,
    date: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
    ending: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
    client: PropTypes.shape({
      email: PropTypes.string,
      name: PropTypes.string,
      surname: PropTypes.string,
      sub: PropTypes.string,
      subscribed: PropTypes.bool
    }),
    facility: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string
    })
  }).isRequired,
  /** type of the Bootstrap panel */
  type: PropTypes.string,
  /** whether the removal button should be hidden or not */
  disabled: PropTypes.bool,
  /** update callback function called on Massage cancellation */
  getCallback: PropTypes.func
};

MyMassagePanel.defaultProps = {
  type: "default",
  disabled: true,
  getCallback: null
};

export default MyMassagePanel;
