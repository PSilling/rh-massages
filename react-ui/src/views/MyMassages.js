// react imports
import React, { Component } from 'react';

// component imports
import CancelButton from '../components/buttons/CancelButton';
import CalendarButton from '../components/iconbuttons/CalendarButton';

// module imports
import moment from 'moment';

// util imports
import Auth from '../util/Auth';
import _t from '../util/Translations';
import Util from '../util/Util';

/**
 * Tabled list of all assigned massages.
 */
class MassagesList extends Component {

  state = {massages: []}

  componentDidMount() {
    Util.clearAllIntervals();

    this.getMassages();

    setInterval(() => {
      this.getMassages();
    }, Util.AUTO_REFRESH_TIME * 60);
  }

  getMassages = () => {
    Util.get(Util.MASSAGES_URL + "client", (json) => {
      json.sort(function(a, b) {
        return a.date - b.date;
      });

      for (var i = 0; i < json.length; i++) {
        if (moment(json[i].ending).isBefore(moment())) {
          json.splice(i, 1);
          i--;
        }
      }

      if (!Util.arraysEqual(this.state.massages, json)) {
        this.setState({massages: json});
      }
    });
  }

  cancelMassage = (massage) => {
    Util.put(Util.MASSAGES_URL + massage.id, {
      date: massage.date,
      ending: massage.ending,
      masseuse: massage.masseuse,
      client: null,
      facility: massage.facility
    }, this.getMassages);
  }

  render() {
    return(
      <div>
        <h1>
          { _t.translate('My Massages') }
        </h1>
        <hr />
        <table className="table table-hover table-responsive table-striped">
          <thead>
            <tr>
              <th>{ _t.translate('Facility') }</th>
              <th>{ _t.translate('Date') }</th>
              <th>{ _t.translate('Time') }</th>
              <th>{ _t.translate('Masseur/Masseuse') }</th>
              <th>{ _t.translate('Event') }</th>
              <th></th>
            </tr>
          </thead>
          {this.state.massages.length > 0 ?
            <tbody>
              {this.state.massages.map((item, index) => (
                <tr key={index}>
                  <td>{item.facility.name}</td>
                  <td>{moment(item.date).format("dd DD. MM.")}</td>
                  <td>{moment(item.date).format("HH:mm") + "â€“" + moment(item.ending).format("HH:mm")}</td>
                  <td>{item.masseuse}</td>
                  <td width="55px">
                    <span className="pull-right">
                      <CalendarButton disabled={false} onAdd={() => Util.addToCalendar(item)} />
                    </span>
                  </td>
                  <td width="55px">
                    <span className="pull-right">
                      <CancelButton onCancel={() => this.cancelMassage(item)}
                        disabled={(moment(item.date).diff(moment(), 'minutes') <= Util.CANCELLATION_LIMIT) && !Auth.isAdmin() ? true : false} />
                    </span>
                  </td>
                </tr>
              ))}
            </tbody>
          : <tbody>
            <tr>
              <th colSpan="6">
                { _t.translate('None') }
              </th>
            </tr>
          </tbody>
        }
        </table>
      </div>
    );
  }
}

const MyMassages = () => (
  <div>
    <MassagesList />
  </div>
);

export default MyMassages
