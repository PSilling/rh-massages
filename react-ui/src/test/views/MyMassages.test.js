// react imports
import React from 'react';
import { MemoryRouter } from 'react-router';
import { Link } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';

// test imports
import MyMassages from '../../views/MyMassages';
import MyMassagePanel from '../../components/panels/MyMassagePanel';
import _t from '../../util/Translations';

// test mocks
jest.mock('../../util/Util');

afterAll(() => {
  jest.resetAllMocks();
});

test('renders content correctly', () => {
  const testRenderer = TestRenderer.create(
          <MemoryRouter>
            <MyMassages />
          </MemoryRouter>
        ),
        testInstance = testRenderer.root;

  let link = testInstance.findByType(Link),
      treeJSON = testRenderer.toJSON();

  expect(treeJSON).toMatchSnapshot();
});

test('properly changes state variables', () => {
  const testMassages = [{
          id: 1,
          date: new Date(),
          ending: new Date(),
          client: null,
          facility: { id: 1, name: "test" }
        }],
        testRenderer = TestRenderer.create(
          <MemoryRouter>
            <MyMassages />
          </MemoryRouter>
        ),
        testInstance = testRenderer.root;

  testMassages[0].ending.setHours(testMassages[0].ending.getHours() + 1);
  let view = testInstance.findByType(MyMassages),
      panels = testInstance.findAllByType(MyMassagePanel);

  expect(panels.length).toBe(0);
  view.instance.setState({ loading: true, massages: testMassages });
  view.instance.getMassages();
  expect(view.instance.state.massages).toEqual([]);
  expect(view.instance.state.loading).toBe(false);

  view.instance.setState({ massages: testMassages });
  let headers = testInstance.findAllByType('h1');
  panels = testInstance.findAllByType(MyMassagePanel);
  expect(headers.length).toBe(2);
  expect(headers[1].props.children).toEqual(_t.translate('Today'));
  expect(panels.length).toBe(1);
  expect(panels[0].props.type).toBe("info");
  expect(panels[0].props.massage).toBe(testMassages[0]);
  expect(panels[0].props.disabled).toBe(false);

  panels[0].props.getCallback();
  expect(view.instance.state.massages).toEqual([]);
});
