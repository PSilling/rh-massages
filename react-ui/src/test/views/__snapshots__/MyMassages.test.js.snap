// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MyMassages />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <InfoAlert
          onClose={[Function]}
>
          Na této stránce můžete vidět své zapsané masáže. Použitím tlačítka kalendáře lze podle dané masáže vygenerovat Google událost.
</InfoAlert>,
        <h1>
          Moje masáže
</h1>,
        <hr />,
        <h3>
          Žádné – 
          <Link
                    replace={false}
                    style={
                              Object {
                                        "color": "#595959",
                                      }
                    }
                    to="/"
          >
                    Jít na masáže
          </Link>
</h3>,
      ],
      "className": "my-3",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Na této stránce můžete vidět své zapsané masáže. Použitím tlačítka kalendáře lze podle dané masáže vygenerovat Google událost.",
          "onClose": [Function],
        },
        "ref": null,
        "rendered": "Na této stránce můžete vidět své zapsané masáže. Použitím tlačítka kalendáře lze podle dané masáže vygenerovat Google událost.",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            false,
            "Moje masáže",
          ],
        },
        "ref": null,
        "rendered": Array [
          false,
          "Moje masáže",
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Žádné – ",
            <Link
              replace={false}
              style={
                            Object {
                                          "color": "#595959",
                                        }
              }
              to="/"
>
              Jít na masáže
</Link>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Žádné – ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Jít na masáže",
              "replace": false,
              "style": Object {
                "color": "#595959",
              },
              "to": "/",
            },
            "ref": null,
            "rendered": "Jít na masáže",
            "type": [Function],
          },
        ],
        "type": "h3",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <InfoAlert
            onClose={[Function]}
>
            Na této stránce můžete vidět své zapsané masáže. Použitím tlačítka kalendáře lze podle dané masáže vygenerovat Google událost.
</InfoAlert>,
          <h1>
            Moje masáže
</h1>,
          <hr />,
          <h3>
            Žádné – 
            <Link
                        replace={false}
                        style={
                                    Object {
                                                "color": "#595959",
                                              }
                        }
                        to="/"
            >
                        Jít na masáže
            </Link>
</h3>,
        ],
        "className": "my-3",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Na této stránce můžete vidět své zapsané masáže. Použitím tlačítka kalendáře lze podle dané masáže vygenerovat Google událost.",
            "onClose": [Function],
          },
          "ref": null,
          "rendered": "Na této stránce můžete vidět své zapsané masáže. Použitím tlačítka kalendáře lze podle dané masáže vygenerovat Google událost.",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              false,
              "Moje masáže",
            ],
          },
          "ref": null,
          "rendered": Array [
            false,
            "Moje masáže",
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Žádné – ",
              <Link
                replace={false}
                style={
                                Object {
                                                "color": "#595959",
                                              }
                }
                to="/"
>
                Jít na masáže
</Link>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Žádné – ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Jít na masáže",
                "replace": false,
                "style": Object {
                  "color": "#595959",
                },
                "to": "/",
              },
              "ref": null,
              "rendered": "Jít na masáže",
              "type": [Function],
            },
          ],
          "type": "h3",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
